### ISSUE TITLE
NaN value in loss when running AMP training with B580

### ISSUE THREAD

**Issue by xiangyang-95** [intel/intel-extension-for-pytorch#814]
### Describe the bug

NaN value in loss when training for a few iterations with AMP. The code is from: [link](https://pytorch.org/docs/stable/notes/get_start_xpu.html#train-with-amp)

```
import torch
import torchvision

LR = 0.001
DOWNLOAD = True
DATA = "datasets/cifar10/"

use_amp=True

transform = torchvision.transforms.Compose(
    [
        torchvision.transforms.Resize((224, 224)),
        torchvision.transforms.ToTensor(),
        torchvision.transforms.Normalize((0.5, 0.5, 0.5), (0.5, 0.5, 0.5)),
    ]
)
train_dataset = torchvision.datasets.CIFAR10(
    root=DATA,
    train=True,
    transform=transform,
    download=DOWNLOAD,
)
train_loader = torch.utils.data.DataLoader(dataset=train_dataset, batch_size=16)
train_len = len(train_loader)

model = torchvision.models.resnet50()
criterion = torch.nn.CrossEntropyLoss()
optimizer = torch.optim.SGD(model.parameters(), lr=LR, momentum=0.9)
scaler = torch.amp.GradScaler(enabled=use_amp)

model.train()
model = model.to("xpu")
criterion = criterion.to("xpu")

print(f"Initiating training")
for batch_idx, (data, target) in enumerate(train_loader):
    data = data.to("xpu")
    target = target.to("xpu")
    # set dtype=torch.bfloat16 for BF16
    with torch.autocast(device_type="xpu", dtype=torch.float16, enabled=use_amp):
        output = model(data)
        loss = criterion(output, target)
    scaler.scale(loss).backward()
    scaler.step(optimizer)
    scaler.update()
    optimizer.zero_grad()
    if (batch_idx + 1) % 10 == 0:
         iteration_loss = loss.item()
         print(f"Iteration [{batch_idx+1}/{train_len}], Loss: {iteration_loss:.4f}")

torch.save(
    {
        "model_state_dict": model.state_dict(),
        "optimizer_state_dict": optimizer.state_dict(),
    },
    "checkpoint.pth",
)

print("Execution finished")
```

### Versions

```
[W424 00:27:56.943348449 OperatorEntry.cpp:154] Warning: Warning only once for all operators,  other operators may also be overridden.
  Overriding a previously registered kernel for the same operator and the same dispatch key
  operator: aten::_validate_compressed_sparse_indices(bool is_crow, Tensor compressed_idx, Tensor plain_idx, int cdim, int dim, int nnz) -> ()
    registered at /pytorch/build/aten/src/ATen/RegisterSchema.cpp:6
  dispatch key: XPU
  previous kernel: registered at /pytorch/build/aten/src/ATen/RegisterCPU.cpp:30477
       new kernel: registered at /build/intel-pytorch-extension/build/Release/csrc/gpu/csrc/aten/generated/ATen/RegisterXPU.cpp:468 (function operator())
Collecting environment information...
=====================================
PyTorch version:   2.6.0+xpu
PyTorch CXX11 ABI: Yes
IPEX version:      2.6.10+xpu
IPEX commit:       b3397a609
Build type:        Release

OS:                Debian GNU/Linux 12 (bookworm) (x86_64)
GCC version:       (Debian 12.2.0-14) 12.2.0
Clang version:     N/A
IGC version:       N/A
CMake version:     N/A
Libc version:      glibc-2.36

Python version:    3.11.2 (main, Nov 30 2024, 21:22:50) [GCC 12.2.0] (64-bit runtime)
Python platform:   Linux-6.11.0-24-generic-x86_64-with-glibc2.36
Is XPU available:  True
DPCPP runtime:     N/A
MKL version:       N/A

GPU models and configuration onboard: 
N/A

GPU models and configuration detected: 
* [0] _XpuDeviceProperties(name='Intel(R) Graphics [0xe20b]', platform_name='Intel(R) oneAPI Unified Runtime over Level-Zero', type='gpu', driver_version='1.6.31294+20', total_memory=11605MB, max_compute_units=160, gpu_eu_count=160, gpu_subslice_count=20, max_work_group_size=1024, max_num_sub_groups=64, sub_group_sizes=[16 32], has_fp16=1, has_fp64=1, has_atomic64=1)

Driver version: 
* intel_opencl: 24.39.31294.20-1032~22.04
* level_zero:   N/A

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Address sizes:                        46 bits physical, 57 bits virtual
Byte Order:                           Little Endian
CPU(s):                               32
On-line CPU(s) list:                  0-31
Vendor ID:                            GenuineIntel
BIOS Vendor ID:                       Intel(R) Corporation
Model name:                           Intel(R) Xeon(R) w5-3435X
BIOS Model name:                      Intel(R) Xeon(R) w5-3435X  CPU @ 3.1GHz
BIOS CPU family:                      179
CPU family:                           6
Model:                                143
Thread(s) per core:                   2
Core(s) per socket:                   16
Socket(s):                            1
Stepping:                             8
CPU(s) scaling MHz:                   31%
CPU max MHz:                          4700.0000
CPU min MHz:                          800.0000
BogoMIPS:                             6192.00
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 cat_l2 cdp_l3 intel_ppin cdp_l2 ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local split_lock_detect user_shstk avx_vnni avx512_bf16 wbnoinvd dtherm ida arat pln pts hwp hwp_act_window hwp_epp hwp_pkg_req vnmi avx512vbmi umip pku ospke waitpkg avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq la57 rdpid bus_lock_detect cldemote movdiri movdir64b enqcmd fsrm md_clear serialize tsxldtrk pconfig arch_lbr ibt amx_bf16 avx512_fp16 amx_tile amx_int8 flush_l1d arch_capabilities
Virtualization:                       VT-x
L1d cache:                            768 KiB (16 instances)
L1i cache:                            512 KiB (16 instances)
L2 cache:                             32 MiB (16 instances)
L3 cache:                             45 MiB (1 instance)
NUMA node(s):                         1
NUMA node0 CPU(s):                    0-31
Vulnerability Gather data sampling:   Not affected
Vulnerability Itlb multihit:          Not affected
Vulnerability L1tf:                   Not affected
Vulnerability Mds:                    Not affected
Vulnerability Meltdown:               Not affected
Vulnerability Mmio stale data:        Not affected
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed:               Not affected
Vulnerability Spec rstack overflow:   Not affected
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; Enhanced / Automatic IBRS; IBPB conditional; RSB filling; PBRSB-eIBRS SW sequence; BHI BHI_DIS_S
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Not affected

Versions of relevant libraries:
[pip] dpcpp-cpp-rt==2025.0.4
[pip] impi-devel==2021.14.1
[pip] impi-rt==2021.14.1
[pip] intel-cmplr-lib-rt==2025.0.4
[pip] intel-cmplr-lib-ur==2025.0.4
[pip] intel-cmplr-lic-rt==2025.0.4
[pip] intel_extension_for_pytorch==2.6.10+xpu
[pip] intel-opencl-rt==2025.0.4
[pip] intel-openmp==2025.0.4
[pip] intel-pti==0.10.0
[pip] intel-sycl-rt==2025.0.4
[pip] mkl==2025.0.1
[pip] mkl-dpcpp==2025.0.1
[pip] numpy==2.2.5
[pip] oneccl==2021.14.1
[pip] oneccl-bind-pt==2.6.0+xpu
[pip] oneccl-devel==2021.14.1
[pip] onemkl-sycl-blas==2025.0.1
[pip] onemkl-sycl-datafitting==2025.0.1
[pip] onemkl-sycl-dft==2025.0.1
[pip] onemkl-sycl-lapack==2025.0.1
[pip] onemkl-sycl-rng==2025.0.1
[pip] onemkl-sycl-sparse==2025.0.1
[pip] onemkl-sycl-stats==2025.0.1
[pip] onemkl-sycl-vm==2025.0.1
[pip] pytorch-triton-xpu==3.2.0
[pip] torch==2.6.0+xpu
[pip] torchao==0.10.0+xpu
[pip] torchaudio==2.6.0+xpu
[pip] torchvision==0.21.0+xpu
[pip] transformers==4.51.3
```

**Comment by jingxu10** [comment]
We will see if we can reproduce the issue.

**Comment by ZhaoqiongZ** [comment]
Hi @xiangyang-95 , please use `GradScaler` with `device="xpu"`
`scaler = torch.amp.GradScaler(device="xpu", enabled=use_amp)` we have updated the proper script to https://pytorch.org/docs/main/notes/get_start_xpu.html
I can not reproduce the loss NaN on my side 
```
Initiating training
Iteration [10/391], Loss: 4.3981
Iteration [20/391], Loss: 2.3955
Iteration [30/391], Loss: 2.2496
Iteration [40/391], Loss: 2.3274
Iteration [50/391], Loss: 2.2701
Iteration [60/391], Loss: 2.2840
Iteration [70/391], Loss: 2.2805
Iteration [80/391], Loss: 2.2699
Iteration [90/391], Loss: 2.1849
Iteration [100/391], Loss: 2.2332
Iteration [110/391], Loss: 2.2170
Iteration [120/391], Loss: 2.1099
Iteration [130/391], Loss: 2.1746
Iteration [140/391], Loss: 2.1728
Iteration [150/391], Loss: 2.1178
Iteration [160/391], Loss: 2.0310
Iteration [170/391], Loss: 2.0614
Iteration [180/391], Loss: 2.1202
Iteration [190/391], Loss: 2.0687
Iteration [200/391], Loss: 2.0429
Iteration [210/391], Loss: 2.0693
Iteration [220/391], Loss: 2.0527
Iteration [230/391], Loss: 1.9904
Iteration [240/391], Loss: 2.0522
Iteration [250/391], Loss: 1.9830
Iteration [260/391], Loss: 1.9489
Iteration [270/391], Loss: 2.0620
Iteration [280/391], Loss: 1.9915
Iteration [290/391], Loss: 1.9265
Iteration [300/391], Loss: 2.0216
Iteration [310/391], Loss: 1.9691
Iteration [320/391], Loss: 2.0686
Iteration [330/391], Loss: 1.9526
Iteration [340/391], Loss: 1.9399
Iteration [350/391], Loss: 1.8964
Iteration [360/391], Loss: 1.9331
Iteration [370/391], Loss: 1.8219
Iteration [380/391], Loss: 1.9387
Iteration [390/391], Loss: 1.8135
Execution finished

```
